### PR TITLE
Use grpc transport in addsvc example client

### PR DESCRIPTION
**Pull request by jupp0r**
This change replaces manual usage of grpc in the addsvc example with the client from transport/grpc #217

### CODE REVIEW

File: examples/addsvc/client/grpc/client.go [hunk 1]
```diff
@@ -4,20 +4,32 @@ import (
...
+	return client{ctx, sumClient.Endpoint(), concatClient.Endpoint(), logger, replies}
 }
 
 type client struct {
 	context.Context
-	pb.AddClient
+	sumEndpoint    endpoint.Endpoint
+	concatEndpoint endpoint.Endpoint
 	log.Logger
+	replies
```
Comment (peterbourgon) [inline examples/addsvc/client/grpc/client.go]: Why is this included? It's never referenced.

File: examples/addsvc/client/grpc/client.go [hunk 2]
```diff
@@ -4,20 +4,32 @@ import (
...
 	"github.com/go-kit/kit/log"
+	grpctransport "github.com/go-kit/kit/transport/grpc"
 )
 
 // New returns an AddService that's backed by the provided ClientConn.
 func New(ctx context.Context, cc *grpc.ClientConn, logger log.Logger) server.AddService {
-	return client{ctx, pb.NewAddClient(cc), logger}
+	var replies replies
+	sumClient := grpctransport.NewClient(cc, "Add", "sum", encodeSumRequest, decodeSumResponse, replies.sumReply)
+	concatClient := grpctransport.NewClient(cc, "Add", "concat", encodeConcatRequest, decodeConcatResponse, replies.concatReply)
```
Comment (peterbourgon) [inline examples/addsvc/client/grpc/client.go]: Unless I'm missing something, you can pass concrete types directly to the NewClients, and remove the `replies` struct...

File: examples/addsvc/client/grpc/client.go [hunk 3]
```diff
@@ -36,27 +48,41 @@ type client struct {
...
-	reply, err := c.AddClient.Sum(c.Context, request)
+	reply, err := c.sumEndpoint(c.Context, request)
 	if err != nil {
 		c.Logger.Log("err", err) // Without an error return parameter, we can't do anything else...
 		return 0
 	}
-	return int(reply.V)
+
+	r, ok := reply.(server.SumResponse)
+	if !ok {
```
Comment (peterbourgon) [inline examples/addsvc/client/grpc/client.go]: I think if this is not true, it's a programmer error, and it's better to just panic. So this (and all other type assertions) can elide the `, ok` check, IMO.

File: examples/addsvc/client/grpc/encode_decode.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+	"github.com/go-kit/kit/examples/addsvc/server"
+	"golang.org/x/net/context"
+)
+
+func encodeSumRequest(ctx context.Context, req interface{}) (interface{}, error) {
+	sumRequest, ok := req.(server.SumRequest)
+
+	if !ok {
+		return nil, errors.New("sum request broken")
+	}
```
Comment (peterbourgon) [inline examples/addsvc/client/grpc/encode_decode.go]: Likewise above, here.

**Comment by peterbourgon** [comment]
Thank you! Apart from those few things, LGTM.

File: examples/addsvc/client/grpc/encode_decode.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+	concatRequest := req.(server.ConcatRequest)
+
+	pbRequest := &pb.ConcatRequest{
+		A: concatRequest.A,
+		B: concatRequest.B,
+	}
+	return pbRequest, nil
+}
+
+func decodeSumResponse(ctx context.Context, req interface{}) (interface{}, error) {
```
Comment (jmank88) [inline examples/addsvc/client/grpc/encode_decode.go]: Should `req` be renamed to `resp` or `reply`?
Same for decodeConcatResponse.

Comment (jupp0r) [inline examples/addsvc/client/grpc/encode_decode.go]: Definitely, copy + paste error

File: examples/addsvc/client/grpc/client.go [hunk 6]
```diff
@@ -4,19 +4,25 @@ import (
...
 	"github.com/go-kit/kit/log"
+	grpctransport "github.com/go-kit/kit/transport/grpc"
 )
 
 // New returns an AddService that's backed by the provided ClientConn.
 func New(ctx context.Context, cc *grpc.ClientConn, logger log.Logger) server.AddService {
-	return client{ctx, pb.NewAddClient(cc), logger}
+	var sumReply, concatReply string
+	sumClient := grpctransport.NewClient(cc, "Add", "sum", encodeSumRequest, decodeSumResponse, sumReply)
+	concatClient := grpctransport.NewClient(cc, "Add", "concat", encodeConcatRequest, decodeConcatResponse, concatReply)
```
Comment (basvanbeek) [inline examples/addsvc/client/grpc/client.go]: sumReply and concatReply should not be strings. You should use `pb.SumReply{}` and `pb.SumRequest{}` from the generated protobuf package.

File: examples/addsvc/client/grpc/client.go [hunk 7]
```diff
@@ -4,19 +4,25 @@ import (
...
+
+	return client{ctx, sumClient.Endpoint(), concatClient.Endpoint(), logger}
 }
 
 type client struct {
 	context.Context
-	pb.AddClient
+	sumEndpoint    endpoint.Endpoint
+	concatEndpoint endpoint.Endpoint
```
Comment (basvanbeek) [inline examples/addsvc/client/grpc/client.go]: i would personally use `sum` and `concat` just like the other transports show.

File: examples/addsvc/client/grpc/client.go [hunk 8]
```diff
@@ -4,19 +4,25 @@ import (
...
 )
 
 // New returns an AddService that's backed by the provided ClientConn.
 func New(ctx context.Context, cc *grpc.ClientConn, logger log.Logger) server.AddService {
-	return client{ctx, pb.NewAddClient(cc), logger}
+	var sumReply, concatReply string
+	sumClient := grpctransport.NewClient(cc, "Add", "sum", encodeSumRequest, decodeSumResponse, sumReply)
+	concatClient := grpctransport.NewClient(cc, "Add", "concat", encodeConcatRequest, decodeConcatResponse, concatReply)
+
+	return client{ctx, sumClient.Endpoint(), concatClient.Endpoint(), logger}
```
Comment (basvanbeek) [inline examples/addsvc/client/grpc/client.go]: I would follow the same coding style like httpjson/client.go and do something like this:
```
return client{
    Context: ctx,
    Logger:  logger,
    sum: grpctransport.NewClient(
        ...,
    ).Endpoint(),
    concat: grpctransport.NewClient(
        ...,
    ).Endpoint(),
}

File: examples/addsvc/client/grpc/client.go (hunk 6 above)
Comment (jupp0r) [inline examples/addsvc/client/grpc/client.go]: I took the usage of strings for the ```Invoke()``` from grpcs unit tests. Digging through the source, it's actually undefined behavior[1]:

> If the struct underlying pb does not match the data in buf, the results can be unpredictable.

[1] https://godoc.org/github.com/golang/protobuf/proto#Unmarshal

**Comment by jupp0r** [comment]
Build is broken because etcd upstream removed support for go 1.4: https://github.com/coreos/etcd/commit/25e47db4163505461ee51242d31342dcfe063405#diff-9c306d74ea02bd4f8619aaa482f8f341

**Comment by peterbourgon** [comment]
Too bad. Can you remove 1.4 from the travis.yml?

**Comment by jupp0r** [comment]
I opened #220 for removing 1.4 builds.

**Comment by jupp0r** [comment]
I addressed all comments so far and squashed the commit.

File: examples/addsvc/client/grpc/encode_decode.go [hunk 9]
```diff
@@ -0,0 +1,45 @@
+package grpc
+
+import (
+	"github.com/go-kit/kit/examples/addsvc/pb"
+	"github.com/go-kit/kit/examples/addsvc/server"
+	"golang.org/x/net/context"
+)
```
Comment (peterbourgon) [inline examples/addsvc/client/grpc/encode_decode.go]:
```go
import (
 	"golang.org/x/net/context"
 
	"github.com/go-kit/kit/examples/addsvc/pb"
 	"github.com/go-kit/kit/examples/addsvc/server"
 )
```

**Comment by peterbourgon** [comment]
Yay! Thank you for the contribution!
